plugin/src/main/java/org/opensearch/sql/plugin/rest/RestPPLQueryAction.java

```diff
@@ -122,138 +96,34 @@ protected Set<String> responseParams() {
-      @Override
-      public void onResponse(ExplainResponse response) {
-        sendResponse(channel, OK, new JsonResponseFormatter<ExplainResponse>(PRETTY) {
-          @Override
-          protected Object buildJsonObject(ExplainResponse response) {
-            return response;
-          }
-        }.format(response));
-      }
-
-      @Override
-      public void onFailure(Exception e) {
-        LOG.error("Error happened during explain", e);
-        sendResponse(channel, INTERNAL_SERVER_ERROR,
-            "Failed to explain the query due to error: " + e.getMessage());
-      }
-    };
-  }
-
-  private ResponseListener<QueryResponse> createListener(RestChannel channel,
-                                                         PPLQueryRequest pplRequest) {
-    Format format = pplRequest.format();
-    ResponseFormatter<QueryResult> formatter;
-    if (format.equals(Format.CSV)) {
-      formatter = new CsvResponseFormatter(pplRequest.sanitize());
-    } else if (format.equals(Format.RAW)) {
-      formatter = new RawResponseFormatter();
-    } else if (format.equals(Format.VIZ)) {
-      formatter = new VisualizationResponseFormatter(pplRequest.style());
-    } else {
-      formatter = new SimpleJsonResponseFormatter(PRETTY);
-    }
-    return new ResponseListener<QueryResponse>() {
-      @Override
-      public void onResponse(QueryResponse response) {
-        sendResponse(channel, OK, formatter.format(new QueryResult(response.getSchema(),
-            response.getResults())));
-      }
-
-      @Override
-      public void onFailure(Exception e) {
-        LOG.error("Error happened during query handling", e);
-        if (isClientError(e)) {
-          Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_CUS).increment();
-          reportError(channel, e, BAD_REQUEST);
-        } else {
-          Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_SYS).increment();
-          reportError(channel, e, SERVICE_UNAVAILABLE);
-        }
-      }
-    };
-  }
-
-  private <T> T doPrivileged(PrivilegedExceptionAction<T> action) {
-    try {
-      return SecurityAccess.doPrivileged(action);
-    } catch (IOException e) {
-      throw new IllegalStateException("Failed to perform privileged action", e);
-    }
-  }
-
-  private void sendResponse(RestChannel channel, RestStatus status, String content) {
-    channel.sendResponse(
-        new BytesRestResponse(status, "application/json; charset=UTF-8", content));
+    TransportPPLQueryRequest transportPPLQueryRequest = new TransportPPLQueryRequest(
+            pplRequest.getRequest(),
+            pplRequest.path(),
+            pplRequest.getFormat(),
+            pplRequest.sanitize(),
+            pplRequest.style(),
+            null);
+
+    return channel -> nodeClient.execute(SQLActions.INSTANCE.getSEND_PPL_QUERY_ACTION_TYPE(),
+            transportPPLQueryRequest, new RestToXContentListener<>(channel){
```

Comment:
new ResponseListener<TransportPPLQueryResponse>() {
      @Override
      public void onResponse(QueryResponse response) {
        sendResponse(channel, OK, formatter.format(new QueryResult(response.getSchema(),
            response.getResults())));
      }

      @Override
      public void onFailure(Exception e) {
        LOG.error("Error happened during query handling", e);
        if (isClientError(e)) {
          Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_CUS).increment();
          reportError(channel, e, BAD_REQUEST);
        } else {
          Metrics.getInstance().getNumericalMetric(MetricName.PPL_FAILED_REQ_COUNT_SYS).increment();
          reportError(channel, e, SERVICE_UNAVAILABLE);
        }
      }
}

Reply:
sure, but I will leave the formatter in the transport layer, and only pass the formatted response string in `onResponse()`